drake/solvers/test/testOptimizationProblem.cpp

```diff
@@ -58,7 +60,7 @@ class SixHumpCamelObjective : public TemplatedDifferentiableFunction<SixHumpCame
 void sixHumpCamel() {
   OptimizationProblem prog;
   auto x = prog.addContinuousVariables(2);
-  auto objective = prog.addCost(make_shared<SixHumpCamelObjective>());
+  auto objective = prog.addCost(SixHumpCamelObjective());
```

Comment:
how does this work exactly?  I see the enable_if convertible_to logic above.  but how does it actually turn the instance the objective into the shared ptr?